### PR TITLE
feat(SM-128) implement client

### PR DESCRIPTION
**Pull request by aheeva-yuliya**

### CODE REVIEW

File: pom.xml
Comment (frisonisland) [inline pom.xml]: NIT: small type on "unboudid"

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [inline src/main/java/com/zextras/SlcwClient.java]: What about making "initialize" public and call new SlcwClient(LdapConnection conn, String baseDn) that you already defined?

**Review by frisonisland — CHANGES_REQUESTED** [review]
I left some comments for now. Still checking because it's quite big.

File: src/main/java/com/zextras/SlcwBean.java
Comment (Polpetta) [inline src/main/java/com/zextras/SlcwBean.java:8]: I'd add Javadoc here

File: src/main/java/com/zextras/SlcwClient.java
Comment (Polpetta) [inline src/main/java/com/zextras/SlcwClient.java:14]: Add `@since` and `@authors`, please refer to https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#examples

File: src/main/java/com/zextras/SlcwClient.java
Comment (Polpetta) [reply src/main/java/com/zextras/SlcwClient.java]: Additionally, it is a bad pattern to have object that its instantiation could fail, it hinders testability too

File: src/main/java/com/zextras/operations/AbstractOperationExecutor.java
Comment (Polpetta) [inline src/main/java/com/zextras/operations/AbstractOperationExecutor.java]: Add Javadoc

File: src/main/java/com/zextras/operations/OperationExecutor.java
Comment (Polpetta) [inline src/main/java/com/zextras/operations/OperationExecutor.java]: Add Javadoc

File: pom.xml
Comment (aheeva-yuliya) [reply pom.xml]: which one is small?

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: i want to make sort of connection factory, but till now i can do this first

File: pom.xml
Comment (frisonisland) [reply pom.xml]: Sorry, I made a "typo" on "type". unboudid should be unboundid

File: src/main/java/com/zextras/persistence/SlcwException.java
Comment (Polpetta) [inline src/main/java/com/zextras/persistence/SlcwException.java:6]: Why haven't you use this exception instead of `RuntimeException` in other places of the codebase?

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: I don't understand, what I mean is that initialize is the factory method. You can make it public and remove the constructor
public SlcwClient(String host, int port, String bindDN, String password, String baseDn)

Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: i mean

i dont know if you saw i just left a todo in the begging of the class to make connection factory or smth 
and open the connection and set all the properties are gonna be a job of another class
we can discuss it

but till now to have smth ready to merge i can do that you suggest

Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: ok, got it!

File: src/main/java/com/zextras/persistence/mapping/SlcwField.java
Comment (Polpetta) [inline src/main/java/com/zextras/persistence/mapping/SlcwField.java]: Extra `*`
```suggestion
 * Class which represents a field of an object.
```

File: src/main/java/com/zextras/transcoders/primitive/ValueTranscoder.java
Comment (Polpetta) [inline src/main/java/com/zextras/transcoders/primitive/ValueTranscoder.java:14]: Missing Javadoc

File: src/main/java/com/zextras/transcoders/reflection/ReflectionTranscoder.java
Comment (Polpetta) [inline src/main/java/com/zextras/transcoders/reflection/ReflectionTranscoder.java:10]: Missing Javadoc

File: src/test/java/com/zextras/Main.java
Comment (Polpetta) [inline src/test/java/com/zextras/Main.java]: I think you can remove this file

File: src/test/java/com/zextras/slcwBeans/User.java
Comment (Polpetta) [inline src/test/java/com/zextras/slcwBeans/User.java]:
```suggestion
public class User implements SlcwBean {
```

File: src/main/java/com/zextras/operations/OperationResult.java
Comment (Polpetta) [inline src/main/java/com/zextras/operations/OperationResult.java]: Override `toString` method instead of implementing a new one (which does the same)
```suggestion
  @Override
  public String toString() {
    return stringRepresentation;
  }
```

File: src/test/java/com/zextras/SlcwClientTests.java
Comment (Polpetta) [inline src/test/java/com/zextras/SlcwClientTests.java]: Please statically import `Assertions` pacakge, so that you only have to write `assertThrows`

**Review by Polpetta — CHANGES_REQUESTED** [review]
I suggested some changes here and there in the codebase. There are also some classes that have very few Javadoc description on the class definition. I'd be more descriptive adding what's that class goal and why it is needed, for instance.

Almost all the tests you made are Integration tests, which have to end up with `ClassNameIT.java`, and not `ClassNameTest.java`, which is reserved for. If tests rely on external dependencies (aka LDAP), then are integration tests. If they do not rely on external dependencies (either they don't need them or they are all mocked), then are unit tests.

File: pom.xml
Comment (aheeva-yuliya) [reply pom.xml]: done!

File: src/main/java/com/zextras/operations/AbstractOperationExecutor.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/operations/AbstractOperationExecutor.java]: done!

File: src/main/java/com/zextras/operations/OperationExecutor.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/operations/OperationExecutor.java]: done!

File: src/main/java/com/zextras/persistence/mapping/SlcwField.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/persistence/mapping/SlcwField.java]: done!

File: src/main/java/com/zextras/transcoders/primitive/ValueTranscoder.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/transcoders/primitive/ValueTranscoder.java:14]: done!

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [inline src/main/java/com/zextras/SlcwClient.java]: I'm checking usage of SlcwClient api as well the SlcwEntry.
About setting dn and filter, do you think we can perform it in the mapper, since also update and add method do the same thing?

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [inline src/main/java/com/zextras/SlcwClient.java]: About SearchScope, I find it odd that it is set in the entry, instead I think it makes more sense to be a property of OperationExecutor, because it is the one that executes the operation to decide if he wants one or more record. SlcwEntity should (almost) be a domain level class.

File: src/main/java/com/zextras/transcoders/reflection/ReflectionTranscoder.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/transcoders/reflection/ReflectionTranscoder.java:10]: done!

File: src/test/java/com/zextras/Main.java
Comment (aheeva-yuliya) [reply src/test/java/com/zextras/Main.java]: done!

File: src/test/java/com/zextras/slcwBeans/User.java
Comment (aheeva-yuliya) [reply src/test/java/com/zextras/slcwBeans/User.java]: done!

File: src/test/java/com/zextras/SlcwClientTests.java
Comment (aheeva-yuliya) [reply src/test/java/com/zextras/SlcwClientTests.java]: done!

File: src/main/java/com/zextras/operations/OperationResult.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/operations/OperationResult.java]: done!

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java:14]: done!

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/ReflectionUtils.java:11]: Utility classes should not have public or default constructor.

File: src/main/java/com/zextras/utils/TranscoderUtils.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/TranscoderUtils.java:9]: Utility classes should not have public or default constructor, please fix other occurrences too.

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: done!

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/ReflectionUtils.java]: var is a restricted identifier, please change it to something else.

File: src/main/java/com/zextras/utils/PropertyBuilder.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/PropertyBuilder.java]: unused type parameter T can be removed here?

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: dn and filter depends on method
add and update same but differs from delete and get
if we want to set it inside mapper we have to provide information about method which calls it
what do you think about it?

File: src/main/java/com/zextras/persistence/SlcwException.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/persistence/SlcwException.java:6]: done!

File: src/main/java/com/zextras/transcoders/primitive/AbstractBinaryValueTranscoder.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/transcoders/primitive/AbstractBinaryValueTranscoder.java]: Use protected modifier here instead of public. (public constructors should not be used in abstract classes, as constructors of abstract classes can only be called in constructors of their subclasses)
please fix other occurrences too.

**Review by keshavbhatt — COMMENTED** [review]
made a few suggestions I'll look into the implementation later.

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/ReflectionUtils.java:11]: i didnt get the point about the constructor
there are no constructors

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/ReflectionUtils.java]: done!

File: src/main/java/com/zextras/utils/PropertyBuilder.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/PropertyBuilder.java]: done!

File: src/main/java/com/zextras/transcoders/primitive/AbstractBinaryValueTranscoder.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/transcoders/primitive/AbstractBinaryValueTranscoder.java]: done!

File: src/main/java/com/zextras/utils/TranscoderUtils.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/TranscoderUtils.java:9]: same, there are no constructors

Comment (keshavbhatt) [reply src/main/java/com/zextras/utils/TranscoderUtils.java:9]: I am referring to the default or implicit public constructor that java adds to classes which do not define at least one explicitly. 
To restrict someone from instantiating such classes, since they only contains static medthods(Utility classes, our case here). 

You can hinder that by adding a private constructor. By adding something like below :)

```
private MyUtilClass() {
    throw new IllegalStateException("Utility class");
}
```

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (keshavbhatt) [reply src/main/java/com/zextras/utils/ReflectionUtils.java:11]: Please check [this](https://github.com/Zextras/slcw/pull/2#discussion_r985626361) comment.

File: src/main/java/com/zextras/SlcwBean.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwBean.java:8]: done!

File: src/main/java/com/zextras/utils/ReflectionUtils.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/ReflectionUtils.java:11]: got it!

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: Do they? Asking because getById delete, add, update all do the same operations:
- map entry
- setFilter
- setDn
setFilter and setDn do not have any specifiic logic that depends on called method

Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: sorry i forgot i moved that logic in operation executor
sorry
they separates here
search(entry.getDn(), SearchScope.ONE, entry.getFilter());
connection.delete(entry.getFilter() + "," + entry.getDn());
before is was in one

Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: Ok, then you can just remove entry.setDn and entry.SetFilter right? 
Also I would rethink a bit dn and filter fields, IMHO:
- baseDn can be a SlcwEntry field, but would make it final. This way you do not risk moving things around by mistake. E.g.: what happens if I get an Entry, then change it's baseDn and then call add?
- filter should be something only SlcwClient understands. So the client is able to generate a filter/request based on given entry fields

Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: i moved entry.setDn and entry.SetFilter to mapper
also baseDN final now

Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: To follow up what we talked about yesterday: avoid side effects like passing an object to a method and changing its state

File: src/main/java/com/zextras/utils/ObjectFactory.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/ObjectFactory.java]: Please replace the following with the proposed everywhere
>"Utility class can not been instantiated."

"Utility class cannot be instantiated"

**Review by keshavbhatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/zextras/utils/ObjectFactory.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/utils/ObjectFactory.java]: "been" denotes somthing that occured in the past, please replace it with "be"

**Review by keshavbhatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/zextras/utils/ObjectFactory.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/utils/ObjectFactory.java]: done!

File: src/main/java/com/zextras/persistence/converters/SlcwConverter.java
Comment (keshavbhatt) [inline src/main/java/com/zextras/persistence/converters/SlcwConverter.java:12]: Helper class! Please add a private constructor to prevent creation of object instances.

**Review by keshavbhatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/zextras/persistence/converters/SlcwConverter.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/persistence/converters/SlcwConverter.java:12]: done!

**Review by keshavbhatt — APPROVED** [review]
Assuming its not the final work, & understanding things will change as we advance this futher, it LGTM till now.

File: pom.xml
Comment (frisonisland) [inline pom.xml:142]: Test scope?

File: src/main/java/com/zextras/SlcwClient.java
Comment (frisonisland) [inline src/main/java/com/zextras/SlcwClient.java]: Remove TODO if we already have the factory.

**Review by frisonisland — CHANGES_REQUESTED** [review]
Just some minor changes

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: i will push it after this merging bcz have to write javadoc

Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: will remove comment till now

Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: Not sure I get it, if you need to write the Javadoc then do it :)

File: pom.xml
Comment (aheeva-yuliya) [reply pom.xml:142]: done!

File: src/main/java/com/zextras/SlcwClient.java
Comment (aheeva-yuliya) [reply src/main/java/com/zextras/SlcwClient.java]: there is no connection factory in this pull
add it or leave like it was?

Comment (frisonisland) [reply src/main/java/com/zextras/SlcwClient.java]: Ok then, leave as is

**Review by frisonisland — APPROVED** [review]

**Review by Polpetta — APPROVED** [review]
Waiting for the connection factory, but atm LGTM

**Pull request merged**
